core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -195,45 +197,56 @@ private[spark] class ProcfsMetricsGetter(procfsDir: String = "/proc/") extends L
         if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
           allMetrics.copy(
             jvmVmemTotal = allMetrics.jvmVmemTotal + vmem,
-            jvmRSSTotal = allMetrics.jvmRSSTotal + (rssMem)
+            jvmRSSTotal = allMetrics.jvmRSSTotal + (rssMem),
+            timeStamp = System.currentTimeMillis
           )
         }
         else if (procInfoSplit(1).toLowerCase(Locale.US).contains("python")) {
           allMetrics.copy(
             pythonVmemTotal = allMetrics.pythonVmemTotal + vmem,
-            pythonRSSTotal = allMetrics.pythonRSSTotal + (rssMem)
+            pythonRSSTotal = allMetrics.pythonRSSTotal + (rssMem),
+            timeStamp = System.currentTimeMillis
           )
         }
         else {
           allMetrics.copy(
             otherVmemTotal = allMetrics.otherVmemTotal + vmem,
-            otherRSSTotal = allMetrics.otherRSSTotal + (rssMem)
+            otherRSSTotal = allMetrics.otherRSSTotal + (rssMem),
+            timeStamp = System.currentTimeMillis
           )
         }
       }
     } catch {
       case f: IOException =>
         logWarning("There was a problem with reading" +
           " the stat file of the process. ", f)
-        ProcfsMetrics(0, 0, 0, 0, 0, 0)
+        ProcfsMetrics(0, 0, 0, 0, 0, 0, System.currentTimeMillis)
     }
   }
 
   private[spark] def computeAllMetrics(): ProcfsMetrics = {
     if (!isAvailable) {
-      return ProcfsMetrics(0, 0, 0, 0, 0, 0)
+      return ProcfsMetrics(0, 0, 0, 0, 0, 0, System.currentTimeMillis)
     }
-    val pids = computeProcessTree
-    var allMetrics = ProcfsMetrics(0, 0, 0, 0, 0, 0)
-    for (p <- pids) {
-      allMetrics = addProcfsMetricsFromOneProcess(allMetrics, p)
-      // if we had an error getting any of the metrics, we don't want to report partial metrics, as
-      // that would be misleading.
-      if (!isAvailable) {
-        return ProcfsMetrics(0, 0, 0, 0, 0, 0)
+    val lastMetricComputation = System.currentTimeMillis() - cachedAllMetric.timeStamp
+    // Check whether we have computed the metrics in the past 1s
+    // ToDo: Should we make this configurable?
+    if(lastMetricComputation > Math.min(1000, HEARTBEAT_INTERVAL_MS)) {
```

Comment:
I think it makes sense to make it configurable, defaulting to 10s maybe

Reply:
Yeah, I also think it makes sense to be configurable. What do you think @squito?

Reply:
Just to add more context about having 1000ms here. The pulling request for Metrics system can't be less than 1 second. So user can configure the caching period using heartbeat interval if they want to cache for less than 1 second. The configuration option can let them to have a cache that is valid for more than 1 second.